pkg/eventqueue/eventqueue.go

```diff
@@ -167,13 +172,19 @@ func (ev *Event) WasCancelled() bool {
 // Enqueue pushes the given event onto the EventQueue. If the queue has been
 // stopped, the Event will not be enqueued, and its cancel channel will be
 // closed, indicating that the Event was not ran. This function may block if
-// the queue is at its capacity for events.
+// the queue is at its capacity for events. If a single Event has Enqueue
+// called on it multiple times asynchronously, there is no guarantee as to
+// which one will return the channel which passes results back to the caller.
 func (q *EventQueue) Enqueue(ev *Event) <-chan interface{} {
-
 	if q.notSafeToAccess() || ev == nil {
 		return nil
 	}
 
+	// Events can only be enqueued once.
+	if atomic.AddInt32(&ev.enqueued, 1) > 1 {
+		return nil
```

Comment:
This needs to be addressed somehow, it can be addressed in a different PR. If the caller of this function is **not** checking for the channel to be nil then doing `<-c` where `c` is a `nil` channel will deadlock forever.

Reply:
Maybe document that the caller must check for `nil`, which is an error condition?

Reply:
Note that `nil` return is done in other error conditions right above. Most likely all callers are already checking for `nil` return due to that.

Reply:
> Note that `nil` return is done in other error conditions right above. Most likely all callers are already checking for `nil` return due to that.

@jrajahalme yeah the problem is that they are not. I've checked.

Reply:
Fixed in all callpaths in latest commit. All call paths now check if the resultant channel is nil and return and / or log an appropriate error.

Reply:
If this is an error case, then please return an error from the function. Golang has a ubiquitous standard for this, if there's an error, you return an error. Implementing functions that return error cases in a non-standard way is inviting future misuse.

Reply:
@joestringer fixed!